File: C++/json_list.cpp
#include <stdexcept>
#include "json_list.hpp"
#include "json_parser.hpp"

JSONList::JSONList(std::string data) {
  this->wrapCharacters['['] = ']';
  this->wrapCharacters['"'] = '"';
  this->populate_entries_from_string(data);
}

JSONObject* JSONList::get_item(int index) {
  return this->entries.at(index);
}

int JSONList::get_entry_count() {
  return this->entries.size();
}

JSONList::~JSONList() {
  std::vector<JSONObject*>::iterator it;
  for (it = this->entries.begin(); it != this->entries.end(); it++)
    delete(*it);
}

bool JSONList::is_delim_char(char c) {
  try {
    char temp = this->wrapCharacters.at(c);
  } 
  catch (const std::out_of_range& oor) {
    return false;
  }
  return true;
}

char JSONList::get_closing_delim_from_stack(std::stack<char>& stack) {
  return this->wrapCharacters.at(stack.top());
}

void JSONList::populate_entries_from_string(std::string str) {
  std::stack<char> delimiter_stack;
  std::string current_string;

  for (int i = 0; i < str.length(); ++i) {
    if (!delimiter_stack.empty() && 
        str[i] == get_closing_delim_from_stack(delimiter_stack))
      delimiter_stack.pop();
    else if (is_delim_char(str[i]))
      delimiter_stack.push(str[i]);

    // Check to see if we have reached a comma
    if ((str[i] == ',' || i == str.length() - 1) && delimiter_stack.empty()) {
      // Check if last item
      if (i == str.length() - 1) current_string.push_back(str[i]);
      this->entries.push_back(JSONParser::get_obj_from_str(current_string));
      current_string.erase(0, std::string::npos);
      continue;
    }
    current_string.push_back(str[i]);
  }
}

File: Java/JSONList.java
import java.util.ArrayList;
import java.util.Stack;
import java.util.Dictionary;
import java.util.Hashtable;
/*
 * jsonFASTParse
 * @author: Brandon Milton
 * http://brandonio21.com
 *
 * A small Java library meant to parse JSON in a quick fashion, given that the
 * end-user is aware of the JSON data schema. Does not support dictionaries.
 */

class JSONList extends JSONObject {
  public ArrayList<JSONObject> entries;
  private  Dictionary<Character, Character> wrapCharacters;

  public JSONList(String data) {
    this.entries = new ArrayList<JSONObject>();
    this.wrapCharacters = new Hashtable<Character, Character>();
    wrapCharacters.put('[', ']');
    wrapCharacters.put('"', '"');
    this.populateEntriesFromString(data);
  }

  public int getEntryCount() {
    return this.entries.size();
  }

  public JSONObject getItem(int index) {
    return this.entries.get(index);
  }

  private void populateEntriesFromString(String str) {
    Stack<Character> delimiterStack = new Stack<Character>();
    StringBuilder currentString = new StringBuilder();

    for (int i = 0; i < str.length(); i++) {
      if (!delimiterStack.empty() && str.charAt(i) == getClosingDelimFromStack(delimiterStack))
        delimiterStack.pop();
      else if (isDelimiterCharacter(str.charAt(i)))
        delimiterStack.push(str.charAt(i));

      // Check to see if we have reached a comma
      if ((str.charAt(i) == ',' || i == str.length() - 1) && delimiterStack.empty()) {
        // Check if last item
        if (i == str.length() - 1) currentString.append(str.charAt(i));
        this.entries.add(JSONParser.getObjectFromString(currentString.toString()));
        currentString.delete(0, currentString.length());
        continue;
      }
      currentString.append(str.charAt(i));
    }
  }

  private boolean isDelimiterCharacter(char c) {
    return this.wrapCharacters.get(c) != null;
  }

  private char getClosingDelimFromStack(Stack<Character> stack) {
    return this.wrapCharacters.get(stack.peek());
  }
}
